### PR TITLE
[SIL][Frontend] Simplify debug info generation flow for SIL files

### PR DESCRIPTION
**Pull request by mshockwave**
- If any of the `-g<kind>` flag is given -- except `-gnone`, debug info will be printed into every generated SIL files.
 - The `-gsil` is deprecated in favor of `-sil-based-debuginfo` to avoid confusion with -g, -gnone, -gline-tables-only etc. The SILDebugInfoGenerator Pass now generates intermediate SIL file with name "<output file>.sil_dbg_<n>.sil". Other functionalities of that Pass remain the same.

For `-sil-based-debuginfo`, I tried to avoid generating intermediate SIL file (i.e. \*.sil_dbg_\*.sil) if the input is already a SIL file -- the idea was to modify SILParser if the input is SIL. But it turned out to take non-trivial amount of changes in SILParser, which is probably not worth it for a feature that is less-frequently used. So I eventually fallback to using the existing round-trip-solution by SILDebugInfoGenerator, which always generates an intermediates SIL file, regardless of the input file type.

This PR should be synchronized with https://github.com/apple/swift-driver/pull/737

### CODE REVIEW

**Comment by mshockwave** [comment]
@swift-ci test

File: lib/Driver/ToolChains.cpp [hunk 1]
```diff
@@ -325,6 +325,12 @@ void ToolChain::addCommonFrontendArgs(const OutputInfo &OI,
     auto OptArg = inputArgs.getLastArgNoClaim(options::OPT_O_Group);
     if (!OptArg || OptArg->getOption().matches(options::OPT_Onone))
       arguments.push_back("-enable-anonymous-context-mangled-names");
+
+    // If the output file format is SIL, we will always print the debug info.
+    if (OI.CompilerOutputType == file_types::ID::TY_SIL) {
+      arguments.push_back("-Xllvm");
+      arguments.push_back("-sil-print-debuginfo");
```
Comment (adrian-prantl) [inline lib/Driver/ToolChains.cpp:332]: Do we also want to rename this option to -g (in a subsequent patch)?

**Review by adrian-prantl — APPROVED** [review]
Thanks, this looks good, and is makes the options less confusion.

File: lib/Driver/ToolChains.cpp (hunk 1 above)
Comment (mshockwave) [reply lib/Driver/ToolChains.cpp:332]: I did think about removing `-sil-print-debuginfo`. For example, adding a new boolean field in `SILOptions` for this purpose and set it up whenever `-g` is given. And the field will later be consumed by `SILPrinter`.
But then we lost the ability to print debug info into SIL file _without_ the help of driver -- for instance, directly using `swift-frontend -frontend ...`.  Also, IIUC, the new swift driver can only communicate with frontend via command line flags (please correct me if I'm wrong).
Or are you suggesting to create a _frontend_ flag named after `-g`?

**Comment by mshockwave** [comment]
Alright, instead of removing `-gsil`, now I keep the flag, make it no-op, and print a warning message regarding the deprecation. Not sure if this is the best solution so feel free to provide some suggestions here.

**Comment by mshockwave** [comment]
@swift-ci test

File: lib/Driver/ToolChains.cpp (hunk 1 above)
Comment (atrick) [reply lib/Driver/ToolChains.cpp:332]: AFAICT `-g` is currently the frontend end flag for "emit debug info". I don't think we should also be passing a `-Xllvm` option.

**Review by eeckstein — APPROVED** [review]

File: lib/Driver/ToolChains.cpp (hunk 1 above)
Comment (mshockwave) [reply lib/Driver/ToolChains.cpp:332]: > AFAICT `-g` is currently the frontend end flag for "emit debug info". I don't think we should also be passing a `-Xllvm` option.

You're right about the `-g` flag. Then I'll just consolidate `-Xllvm -sil-print-debuginfo` into `-g` as suggested.

**Comment by mshockwave** [comment]
I just simplified the flags passed from driver: No extra `-Xllvm -sil-print-debuginfo` flags will be passed from driver anymore if `-g` is given. Instead, if any of the debugger level is given, the frontend will populate the newly-added `SILOptions::PrintDebugInfo` flag, which will later be used by `SILPrinter` to print debug info into SIL files.

**Comment by mshockwave** [comment]
@swift-ci test

File: docs/DebuggingTheCompiler.md [hunk 2]
```diff
@@ -25,7 +25,7 @@ benefit of all Swift developers.
         - [Getting CommandLine for swift stdlib from Ninja to enable dumping stdlib SIL](#getting-commandline-for-swift-stdlib-from-ninja-to-enable-dumping-stdlib-sil)
         - [Dumping the SIL and other Data in LLDB](#dumping-the-sil-and-other-data-in-lldb)
     - [Debugging and Profiling on SIL level](#debugging-and-profiling-on-sil-level)
-        - [SIL source level profiling using -gsil](#sil-source-level-profiling-using--gsil)
+        - [SIL source level profiling using -sile-based-debuginfo](#sil-source-level-profiling)
```
Comment (gottesmm) [inline docs/DebuggingTheCompiler.md:28]: @mshockwave Typo.

Comment (gottesmm) [reply docs/DebuggingTheCompiler.md:28]: Feel free to fix in a subsequent commit though.

**Comment by mshockwave** [comment]
Fix test failure in `SILOptimizer/outliner.swift` by updating the CHECK directive to consider the debug location.
@swift-ci test

**Pull request merged**